[CodeGen][NewPM] Port RegAllocPriorityAdvisor analysis to NPM

optimisan opened the pull request:
Similar to #117309.

The advisor and logger are accessed through the provider, which is served by the new PM. Legacy PM forwards calls to the provider.
New PM is a machine function analysis that lazily initializes the provider.

optimisan committed:
[CodeGen][NewPM] Port RegAllocPriorityAdvisor analysis to NPM

optimisan commented:
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/118462?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#118462** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/118462?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/118462?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#117309** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/117309?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* `main`

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

arsenm commented on llvm/lib/CodeGen/RegAllocPriorityAdvisor.cpp:
```diff
@@ -20,76 +20,138 @@
+  switch (Mode) {
+  case RegAllocPriorityAdvisorProvider::AdvisorMode::Default:
+    Provider.reset(
+        new DefaultPriorityAdvisorProvider(/*NotAsRequested*/ false, Ctx));
```
```suggestion
        new DefaultPriorityAdvisorProvider(/*NotAsRequested=*/ false, Ctx));
```

arsenm commented on llvm/include/llvm/CodeGen/RegAllocPriorityAdvisor.h:
```diff
@@ -56,20 +58,81 @@ class DefaultPriorityAdvisor : public RegAllocPriorityAdvisor {
+  virtual ~RegAllocPriorityAdvisorProvider() = default;
+
+  virtual void logRewardIfNeeded(const MachineFunction &MF,
+                                 llvm::function_ref<float()> GetReward) {};
```
Don't need llvm::

arsenm reviewed (commented):
Can you add a test? Does enough of the pipeline work to use a stop-after line in one of the existing tests?

cdevadas commented:
> Can you add a test? Does enough of the pipeline work to use a stop-after line in one of the existing tests?

Not really. A few more passes to be ported before we can use stop-after regalloc in an existing lit test.

optimisan committed:
use the provider as the analysis result

Avoid setting the advisor from the legacy wrapper after setting all
other analyses.

arsenm commented on llvm/include/llvm/CodeGen/RegAllocPriorityAdvisor.h:
```diff
@@ -56,20 +58,80 @@ class DefaultPriorityAdvisor : public RegAllocPriorityAdvisor {
+  virtual ~RegAllocPriorityAdvisorProvider() = default;
+
+  virtual void logRewardIfNeeded(const MachineFunction &MF,
+                                 llvm::function_ref<float()> GetReward) {};
```
```suggestion
                                 function_ref<float()> GetReward) {};
```

arsenm commented on llvm/lib/CodeGen/MLRegAllocPriorityAdvisor.cpp:
```diff
@@ -150,12 +138,48 @@ class ReleaseModePriorityAdvisorAnalysis final
+    return false;
+  }
+
+  // std::unique_ptr<ReleaseModePriorityAdvisorProvider> Provider;
```
Commented out code

arsenm commented on llvm/lib/CodeGen/MLRegAllocPriorityAdvisor.cpp:
```diff
@@ -254,7 +249,22 @@ class DevelopmentModePriorityAdvisorAnalysis final
+    // contexts. At this point, it'd be an error
+    if (Log->currentContext() != MF.getName()) {
+      MF.getFunction().getContext().emitError(
+          "The training log context shouldn't have had changed.");
```
Errors should start with lowercase and should not use contractions

arsenm commented on llvm/lib/CodeGen/RegAllocPriorityAdvisor.cpp:
```diff
@@ -20,76 +20,138 @@
+  }
+  if (!Provider)
+    Provider.reset(
+        new DefaultPriorityAdvisorProvider(/*NotAsRequested*/ true, Ctx));
```
```suggestion
        new DefaultPriorityAdvisorProvider(/*NotAsRequested=*/ true, Ctx));
```
This "NotAsRequested" name is bad, I don't know what it means

optimisan replied:
This is being used only to print an error for falling back to the default allocator. Can remove this param and print it here.

arsenm commented on llvm/include/llvm/CodeGen/RegAllocPriorityAdvisor.h:
```diff
@@ -56,20 +58,80 @@ class DefaultPriorityAdvisor : public RegAllocPriorityAdvisor {
   virtual std::unique_ptr<RegAllocPriorityAdvisor>
   getAdvisor(const MachineFunction &MF, const RAGreedy &RA) = 0;
+
+  void setAnalyses(SlotIndexes *SI) { this->SI = SI; }
```
Why can't this just be set at construction time?

optimisan replied:
This provider is initialized only once for the pipeline and we just set the new analysis results while running Greedy.

arsenm replied:
OK, so this is per-function state stored in a module analysis? That doesn't sound OK. The current RegAllocPriorityAdvisorAnalysis is an ImmutablePass, and has none of the function state. getAdvisor produces the per-function instance for the run of the pass. I think the signature of getAdvisor needs to gain parameters for all the function analyses passed through during the greedy run

optimisan replied:
Okay, removing setAnalyses would make it simpler as well.

optimisan committed:
Suggestions and remove createProvider methods

optimisan committed:
Apply suggestions and remove setAnalyses

optimisan committed:
run clang format

optimisan committed:
Add newline at EOF

arsenm commented on llvm/lib/CodeGen/RegAllocPriorityAdvisor.cpp:
```diff
@@ -20,107 +20,172 @@
+    Provider.reset(
+        new DefaultPriorityAdvisorProvider(/*NotAsRequested=*/false, Ctx));
+  else
+    initializeMLProvider(Mode, Ctx);
```
I don't understand initializeMLProvider, it's just splitting another set new instantiation. Each one of these should be directly handled as a switch over the modes

optimisan committed:
make SI a reference

optimisan committed:
AS, remove initializeMLProvider()

arsenm commented on llvm/lib/CodeGen/RegAllocPriorityAdvisor.cpp:
```diff
@@ -20,107 +20,181 @@
+    Provider.reset(
+        new DefaultPriorityAdvisorProvider(/*NotAsRequested=*/true, Ctx));
+#endif
+    assert(Provider && "PriorityAdvisorProvider cannot be null");
```
Unreachable assert

optimisan replied:
Right

optimisan committed:
remove assert

arsenm commented on llvm/lib/CodeGen/RegAllocPriorityAdvisor.cpp:
```diff
@@ -20,107 +20,181 @@
+    Provider.reset(
+        new DefaultPriorityAdvisorProvider(/*NotAsRequested=*/true, Ctx));
+#endif
+    assert(Provider && "PriorityAdvisorProvider cannot be null");
```
```suggestion
```
